### PR TITLE
APPS-1191 optional tool input undeclared in scatter

### PR DESCRIPTION
**Pull request by YuxinShi0423**

### CODE REVIEW

**Review by commandlinegirl — DISMISSED** [review]
This LGTM but is it possible to add a test for this? If we don't have a mechanism yet for running our own CWL scripts in integration tests, could you add a minimal test script somewhere anyway, so we can add it when test suite refactor is done?

File: executorCommon/src/main/scala/dx/executor/WorkflowExecutor.scala [hunk 1]
```diff
@@ -721,7 +721,7 @@ abstract class WorkflowExecutor[B <: Block[B]](jobMeta: JobMeta, separateOutputs
     def launch(): Map[DxName, ParameterLink] = {
       val outputs: Map[DxName, ParameterLink] = block.kind match {
         case BlockKind.CallWithSubexpressions | BlockKind.CallFragment =>
-          launchCall(block.index)
+          launchBlockCall(block.index)
```
Comment (YuxinShi0423) [inline executorCommon/src/main/scala/dx/executor/WorkflowExecutor.scala:724]: Rename this function here to fit the three launch scenarios: block, conditional, scatter.

All three would later call `launchStepCall` function to call the applet and launch a job.

File: executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala [hunk 2]
```diff
@@ -446,13 +446,14 @@ case class CwlWorkflowExecutor(workflow: Workflow,
...
-        isConditinal: Boolean = false
+        inputs: Option[Map[DxName, (WorkflowStepInput, (CwlType, CwlValue))]] = None,
+        isConditional: Boolean = false
     ): Map[DxName, (WorkflowStepInput, (CwlType, CwlValue))] = {
       // Evaluate all the step inputs for a call. There may be step inputs that are not passed to
       // the callee but are referred to in expressions of other step inputs' valueFrom fields. Note
       // that the step input's type here might not match the callee's input type if valueFrom is
       // specified.
-      val stepInputs = step.inputs.map(evaluateStepInput(_, cwlEnv)).toMap
+      val stepInputs = inputs.getOrElse(step.inputs.map(evaluateStepInput(_, cwlEnv)).toMap)
```
Comment (YuxinShi0423) [inline executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala:456]: For scatter, the actual `stepInputs` is first evaluated as `allStepInputs`. https://github.com/dnanexus/dxCompiler/blob/1d0c55a4a624e84b35d552c9107b3fe9a61d4475/executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala#L752-L758
So add an optional arg here to override.

File: executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala [hunk 3]
```diff
@@ -749,51 +756,11 @@ case class CwlWorkflowExecutor(workflow: Workflow,
               case (dxName, tv) => dxName -> (stepInputs(dxName)._1, tv)
             }
             .toMap
-        lazy val evalInputs: ObjectValue = createEvalInputs(allStepInputs.values.toMap)
-        val scatterJobInputs = runInputs.map {
```
Comment (YuxinShi0423) [inline executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala:753]: Here is where the issue came from. It should not look for `runInputs` in current `allStepInputs` map. Instead step inputs should be evaluated and then checked when passed to the callee in the `launchStepCall`.

https://github.com/dnanexus/dxCompiler/blob/1d0c55a4a624e84b35d552c9107b3fe9a61d4475/executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala#L606-L624

File: executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala [hunk 4]
```diff
@@ -602,8 +603,9 @@ case class CwlWorkflowExecutor(workflow: Workflow,
       (dxExecution, executableLink, execName)
     }
 
-    private def launchCall(stepInputs: Map[DxName, (WorkflowStepInput, (CwlType, CwlValue))],
-                           blockIndex: Int): Map[DxName, ParameterLink] = {
+    private def launchStepCall(stepInputs: Map[DxName, (WorkflowStepInput, (CwlType, CwlValue))],
```
Comment (YuxinShi0423) [inline executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala:606]: Portal function that accepts evaluated step inputs from block/conditional/scatter and checked against target callee's inputs, then `launchCall`.

File: compiler/src/test/resources/bugs/apps-1191_scatter_optional_input.json [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            "type": [
+              "null",
+              "string"
+            ],
+            "id": "undeclared_with_default",
+            "inputBinding": {
+              "position": 2,
+              "prefix": "\nthis should appear:"
+            },
+            "default": "there is a default."
```
Comment (YuxinShi0423) [inline compiler/src/test/resources/bugs/apps-1191_scatter_optional_input.json]: These two task inputs are not declared in wf. However, the one with default will be added to the output and the one without default in position 1 will be skipped.

So the final scatter output is:
```
[[null," scatter number: 1 this should appear: there is a default.\n",null," scatter number: 1 this should appear: there is a default.\n"],[null," scatter number: 2 this should appear: there is a default.\n",null," scatter number: 2 this should appear: there is a default.\n"]]
```
https://staging.dnanexus.com/projects/Fy9QqgQ0yzZbg9KXKP4Jz6Yq/monitor/analysis/G9kX7GQ0yzZvKg77BP1YxxXv

File: executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala [hunk 6]
```diff
@@ -481,7 +482,7 @@ case class CwlWorkflowExecutor(workflow: Workflow,
             val cwlType = runInputs
               .get(dxName)
               .map {
-                case inp if isConditinal && CwlOptional.isOptional(sourceType) =>
+                case inp if isConditional && CwlOptional.isOptional(sourceType) =>
```
Comment (commandlinegirl) [inline executorCwl/src/main/scala/dx/executor/cwl/CwlWorkflowExecutor.scala:485]: Thanks for fixing the typos as well!

**Review by commandlinegirl — DISMISSED** [review]

**Comment by YuxinShi0423** [comment]
@commandlinegirl  We actually don't have the test `CwlWorkflowExecutorTest`. I was working on adding one, but that needs more time and maybe is out of the scope of this PR. I will create a ticket for that if we want. Thanks!

**Comment by commandlinegirl** [comment]
> @commandlinegirl We actually don't have the test `CwlWorkflowExecutorTest`. I was working on adding one, but that needs more time and maybe is out of the scope of this PR. I will create a ticket for that if we want. Thanks!

Thanks for the ticket! Let's add it separately.

**Review by commandlinegirl — APPROVED** [review]

**Comment by YuxinShi0423** [comment]
Created ticket https://jira.internal.dnanexus.com/browse/APPS-1195. I am running `cwl_cromwell` test suite and if it goes well I will merge this PR. Thanks!

> > @commandlinegirl We actually don't have the test `CwlWorkflowExecutorTest`. I was working on adding one, but that needs more time and maybe is out of the scope of this PR. I will create a ticket for that if we want. Thanks!
> 
> Thanks for the ticket! Let's add it separately.

**Comment by YuxinShi0423** [comment]
All tests in `cwl_cromwell` passed. Thanks! @commandlinegirl

**Comment by commandlinegirl** [comment]
> All tests in `cwl_cromwell` passed. Thanks! @commandlinegirl

Great! Thanks @YuxinShi0423

**Pull request merged**